### ISSUE TITLE
Fatal error: unexpected enum case 'SKErrorCode(rawValue: 110)'

### ISSUE THREAD

**Issue by karaleuski** [bizz84/SwiftyStoreKit#415]
### Platform

- [ ] iOS

### In app purchase type

- [ ] Non-consumable 

### Environment

- [ ] Production

### Version
SwiftyStoreKit (0.13.1)

### Issue summary
When the Internet is turned off, the application crashes after a purchase request. Stop point: case .cloudServiceRevoked:

Error: Thread 1: Fatal error: unexpected enum case 'SKErrorCode(rawValue: 110)'

The error handler should work and call the appropriate alert.

Thanks

**Comment by Jerland2** [comment]
Also seeing this issue in production, very unfortunate, have you found/made a fix?

**Comment by Jerland2** [comment]
> > Also seeing this issue in production, very unfortunate, have you found/made a fix?
> 
> Hi,
> I could not fix the error (

One workaround is implementing a reachability class and checking that the user is connected before using this call in SwiffyStoreKit.  Stilll needs to be fixed in the lib, but its a fix to fatal production crashes.

**Comment by karaleuski** [comment]
> > > Also seeing this issue in production, very unfortunate, have you found/made a fix?
> > 
> > 
> > Hi,
> > I could not fix the error (
> 
> One workaround is implementing a reachability class and checking that the user is connected before using this call in SwiffyStoreKit. Stilll needs to be fixed in the lib, but its a fix to fatal production crashes.

Logically, the creator of the SwiftyStoreKit should deal with correcting this serious error. :) I will wait for a solution to the problem in future versions.

**Comment by Jerland2** [comment]
> > > > Also seeing this issue in production, very unfortunate, have you found/made a fix?
> > > 
> > > 
> > > Hi,
> > > I could not fix the error (
> > 
> > 
> > One workaround is implementing a reachability class and checking that the user is connected before using this call in SwiffyStoreKit. Stilll needs to be fixed in the lib, but its a fix to fatal production crashes.
> 
> Logically, the creator of the SwiftyStoreKit should deal with correcting this serious error. :) I will wait for a solution to the problem in future versions.

I mean yes it should be fixed, but if you want to stop YOUR users from crashing... that will fix your issue

**Comment by karaleuski** [comment]
> > > > > Also seeing this issue in production, very unfortunate, have you found/made a fix?
> > > > 
> > > > 
> > > > Hi,
> > > > I could not fix the error (
> > > 
> > > 
> > > One workaround is implementing a reachability class and checking that the user is connected before using this call in SwiffyStoreKit. Stilll needs to be fixed in the lib, but its a fix to fatal production crashes.
> > 
> > 
> > Logically, the creator of the SwiftyStoreKit should deal with correcting this serious error. :) I will wait for a solution to the problem in future versions.
> 
> I mean yes it should be fixed, but if you want to stop YOUR users from crashing... that will fix your issue

I am not very correct and fully expressed. Yes, I will definitely try the proposed solution to the problem. Thank!

**Comment by EvsenevDev** [comment]
Hi, waiting for fix, current solution is in **switch**:
`default: print(error.localizedDescription)`

**Comment by astrokin** [comment]
```
extension SKError {
  
  var failedReason: String {
    var failedReason: String!
    switch self.code {
    case .unknown: failedReason = StoreService.Strings.unknownError
    case .storeProductNotAvailable: failedReason = StoreService.Strings.invalidProductId
    case .paymentNotAllowed: failedReason = StoreService.Strings.paymentsNotAllowed
    case .paymentInvalid: failedReason = StoreService.Strings.paymentInvalid
    case .paymentCancelled: failedReason = StoreService.Strings.paymentCancelled
    case .cloudServiceRevoked: failedReason = StoreService.Strings.cloudServicePermissionDenied
    case .cloudServicePermissionDenied: failedReason = StoreService.Strings.cloudServicePermissionDenied
    case .cloudServiceNetworkConnectionFailed: failedReason = StoreService.Strings.cloudServiceNetworkConnectionFailed
    case .clientInvalid: failedReason = StoreService.Strings.clientInvalid
    default:
        failedReason = StoreService.Strings.unknownError
    }
    return failedReason
  }
}
```

**Comment by iKisliy** [comment]
@astrokin where should it be placed? I have crash on completion(.error(error: SKError(_nsError: error as NSError))), I rarely think it's will fix it

**Comment by astrokin** [comment]
@iKisliy yes. that's the fix. Just add such extension and use failedReason for any SKError switches

**Comment by KevinQuisquater** [comment]
Users have experienced that crash in 2 of my apps. Thanks for the fix.
This should maybe be added to the sample code in the documentation on Github and on Cocoapods.org while it is being fixed?

**Comment by astrokin** [comment]
Take into consideration author's note
```
Note from the Author

I started Coding with Flutter, a YouTube video series on building apps with
Flutter. Interested? Subscribe here.
```
I'd like to say that we should care about it by ourselves

**Comment by bizz84** [comment]
Tested demo app with the fix from #423. It no longer crashes when purchasing a product while offline. 

One note about how error handling is implemented in the library. There are various lines like this:
```
completion(.error(error: SKError(_nsError: error as NSError)))
```
The upside of this is that the completion handlers will return a `SKError`, which makes it easier to switch on StoreKit errors. But this requires the workaround in #423 to prevent crashes.

The alternative would be for SwiftyStoreKit to return a regular `Error`, meaning that the client has to convert it to SKError via `SKError(_nsError: error as NSError)` to get StoreKit-specific error codes.

Any opinions?
